packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol

```diff
@@ -91,13 +83,14 @@ library LibPosition {
      * @return rightIndex_ The deepest, right most index relative to the `position`.
      */
     function rightIndex(Position _position, uint256 _maxDepth) internal pure returns (uint64 rightIndex_) {
+        uint256 msb = depth(_position);
         assembly {
-            rightIndex_ := shr(0xC0, shl(0xC0, _position))
-
-            // Walk down to the max depth by moving right
-            for { let i := shr(0x40, _position) } lt(i, sub(_maxDepth, 0x01)) { i := add(i, 0x01) } {
-                rightIndex_ := add(0x01, shl(0x01, rightIndex_))
+            let descent := sub(_maxDepth, msb)
+            for { let i := 0 } lt(i, descent) { i := add(i, 1) } {
```

Comment:
not very happy with the looping in this implementation. But I couldn't figure out a more efficient way.

Reply:
Yeah, good question.. we can get rid of the loop with this:
```solidity
    function rightIndex(Position _position, uint256 _maxDepth) internal pure returns (uint64 rightIndex_) {
        uint256 msb = depth(_position);
        assembly {
            switch eq(msb, _maxDepth)
            case true {
                rightIndex_ := _position
            }
            default {
                let remaining := sub(_maxDepth, msb)
                rightIndex_ := or(shl(remaining, _position), sub(shl(remaining, 1), 1))
            }
        }
        rightIndex_ = indexAtDepth(Position.wrap(rightIndex_));
    }
```
but I also don't like how we have to resort to calling `indexAtDepth` again at the bottom. Do we want the gindex, or the trace index (index at depth) here? If we want the gindex, we can delete the final line of this func.

Reply:
Thank you, this is what I wanted.
We actually can avoid calling `indexAtDepth` and thus avoid running another bitscan on the `rightIndex_`. We know that `rightIndex_` is at max depth right after the switch statement. All that's left is to clear the `_maxDepth` bit to get the actual index.

In any case, the gindex should suffice if `rightIndex` will only be used during resolution.